cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/TransformExecutorFactory.java

```diff
@@ -77,23 +85,88 @@ protected abstract TrackedTransform getTransformation(String pluginType, String
    */
-  public TransformExecutor<T> create(PipelinePhase pipeline) throws Exception {
-    Map<String, TransformDetail> transformations = new HashMap<>();
+  public <KEY_OUT, VAL_OUT> BatchTransformExecutor<T> create(PipelinePhase pipeline,
+                                                             OutputWriter<KEY_OUT, VAL_OUT> outputWriter,
+                                                             Map<String, ErrorOutputWriter<Object, Object>>
+                                                               transformErrorSinkMap)
+    throws Exception {
+    Map<String, BatchTransformDetail> transformations = new HashMap<>();
+    Set<String> sources = pipeline.getSources();
+
+    // Set input and output schema for this stage
     for (String pluginType : pipeline.getPluginTypes()) {
       for (StageInfo stageInfo : pipeline.getStagesOfType(pluginType)) {
         String stageName = stageInfo.getName();
-        outputSchema = stageInfo.getOutputSchema();
+        outputSchemas.put(stageName, stageInfo.getOutputSchema());
         perStageInputSchemas.put(stageName, stageInfo.getInputSchemas());
-        // Wrap each transformation so that each stage is emitting stageName along with the record
-        transformations.put(stageName,
-                            new TransformDetail(getTransformation(pluginType, stageName),
-                                                pipeline.getStageOutputs(stageName)));
       }
     }
 
+    // recursively set BatchTransformDetail for all the stages
+    for (String source : sources) {
+      setBatchTransformDetail(pipeline, source, transformations, transformErrorSinkMap, outputWriter);
+    }
+
     // sourceStageName will be null in reducers, so need to handle that case
     Set<String> startingPoints = (sourceStageName == null) ? pipeline.getSources() : Sets.newHashSet(sourceStageName);
-    return new TransformExecutor<>(transformations, startingPoints);
+    return new BatchTransformExecutor<>(transformations, startingPoints);
+  }
+
+  private <KEY_OUT, VAL_OUT> void setBatchTransformDetail(PipelinePhase pipeline, String stageName,
+                                                          Map<String, BatchTransformDetail> transformations,
+                                                          Map<String, ErrorOutputWriter<Object, Object>>
+                                                            transformErrorSinkMap,
+                                                          OutputWriter<KEY_OUT, VAL_OUT> outputWriter)
+    throws Exception {
+    if (pipeline.getSinks().contains(stageName)) {
+      String pluginType = pipeline.getStage(stageName).getPluginType();
+      // If there is a connector sink/ joiner at the end of pipeline, do not remove stage name. This is needed to save
+      // stageName along with the record in connector sink and joiner takes input along with stageName
+      boolean removeStageName = !(pluginType.equalsIgnoreCase(Constants.CONNECTOR_TYPE) ||
```

Comment:
shouldn't need to ignore case. Technically, plugin types with different casing are different types, though we would never do that on purpose.